### ISSUE TITLE
question on KL computation and lastgaelam

### ISSUE THREAD

**Issue by DaehanKim** [lvwerra/trl#121]
Hi! I have 2 questions on `PPOTrainer.loss`. 

(1) What is purpose of `lastgaelam` when `delta` is actually an advantage estimate? [line link](https://github.com/lvwerra/trl/blob/main/trl/trainer/ppo_trainer.py#L689)

It seems `lastgaelam` is collecting future trajactory advantages at a timestep with decaying factor `self.config.lam`.
Why is this value used instead of textbook advantage `delta`?

(2) Why is KL not computed with expectation over $\pi_\text{old}$ but just with $1/T$? [line link](https://github.com/lvwerra/trl/blob/main/trl/trainer/ppo_trainer.py#L734)

when `logprob` and `oldlogprob` is logprobs with (batch_size, T) shape, I think $KL(\pi_\text{old} | \pi)$ should be `torch.dot(oldlogprob-logprob, F.softmax(oldlogprob, dim=1))`. Why is it computed just as a mean over sequence length? Also I don't know why there are two different KLs(`approxkl`, `policykl`). 

I'd really appreciate if you answer these questions. Thanks!

**Comment by natolambert** [comment]
I would guess most of these questions are answered by the fact that the repo was built to closely mirror the "learning to summarize" work.  Which of these do you think is more impactful in investigating?

**Comment by DaehanKim** [comment]
Thank you for answering! I'll check out the original implementation. 
[this post](https://iclr-blog-track.github.io/2022/03/25/ppo-implementation-details/) contains answers many of my question on their implementations!

**Comment by lvwerra** [comment]
Closing this for now - feel free to re-open if you have any follow-up :)
